src/click/core.py

```diff
@@ -2941,12 +2996,18 @@ def prompt_for_value(self, ctx: Context) -> t.Any:
         """
         assert self.prompt is not None
 
-        # Calculate the default before prompting anything to be stable.
+        # Calculate the default before prompting anything to lock in the value before
+        # attempting any user interaction.
         default = self.get_default(ctx)
 
-        # If this is a prompt for a flag we need to handle this
-        # differently.
+        # A boolean flag can use a simplified [y/n] confirmation prompt.
         if self.is_bool_flag:
+            # If we have no boolean default, we force the user to explicitly provide
+            # one.
+            if default in (UNSET, None):
+                default = None
+            else:
+                default = bool(default)
```

Comment:
When would this be the case? They shouldn't be entering a non-bool default for a bool flag.